proposals/params-span.md

```diff
@@ -0,0 +1,138 @@
+
+For overloads that are applicable in _expanded_ form, [Better function member](https://github.com/dotnet/csharplang/blob/main/spec/expressions.md#better-function-member) will be updated to prefer `params` types in a specific order:
+
+> When performing this evaluation, if `Mp` or `Mq` is applicable in its expanded form, then `Px` or `Qx` refers to a parameter in the expanded form of the parameter list.
+> 
+> In case the parameter type sequences `{P1, P2, ..., Pn}` and `{Q1, Q2, ..., Qn}` are equivalent (i.e. each `Pi` has an identity conversion to the corresponding `Qi`), the following tie-breaking rules are applied, in order, to determine the better function member.
+> 
+> *  If `Mp` is a non-generic method and `Mq` is a generic method, then `Mp` is better than `Mq`.
+> *  ...
+> *  **Otherwise, if both methods have `params` parameters and are applicable only in their expanded forms, and the `params` types are distinct types with equivalent element type (there is an identity conversion between element types), the more specific `params` type is the first of:**
+>    *  **`ReadOnlySpan<T>`**
+>    *  **`Span<T>`**
+>    *  **`T[]`**
+>    *  **`IEnumerable<T>`**
+> *  Otherwise if one member is a non-lifted operator and  the other is a lifted operator, the non-lifted one is better.
+> *  Otherwise, neither function member is better.
+
+### Implicit allocation of arrays
+Array creation expressions that are target-typed to `ReadOnlySpan<T>` or `Span<T>` will be created on the stack if the length of the array is a constant value no more than 8 (an arbitrary limit).
+Otherwise the array will be allocated on the heap.
+
+```csharp
+Span<int> s = new[] { i, j, k };   // int[] on the stack
+WriteLine(fmt, new[] { x, y, z }); // object[] on the stack for WriteLine(string fmt, ReadOnlySpan<object> args);
+```
+
+### Array re-use
+The compiler _may_ reuse an implicitly allocated array across multiple uses within a single thread executing a method:
+- At the same call-site (within a loop) or
+- At distinct call-sites if the lifetime of the spans do not overlap, and the array length is sufficient, and
+  - the element types are managed types that are considered identical by the runtime, or
+  - the element types are unmanaged types of the same size.
+
+An implicitly allocated array may be reused regardless of whether the array was created on the stack or the heap.
+
+### Lowering implicit allocation
+For an array `T[]` of constant length `N`, the compiler will synthesize a `struct` with `N` fields of type `T` where the layout and alignment of the fields matches the alignment of elements in `T[]`.
+The array will be represented with a value of the synthesized `struct`, and a `Span<T>` created from a `ref` to the first field.
+
+For example, a call to `Console.WriteLine(fmt, x, y, z);` would be emitted as:
+```csharp
+[StructLayout(LayoutKind.Sequential)]
+internal struct __ValueArray3<T> { public T Item1, Item2, Item3; };
```

Comment:
May be worth considering the kind of struct we already generate for [fixed size buffers](https://github.com/dotnet/csharplang/blob/main/spec/unsafe-code.md#fixed-size-buffers). Such struct has a single field with the type of an item, but the layout is stretched out. [sharplab](https://sharplab.io/#v2:EYLgZgpghgLgrgJwgZwLQBsCWyaoA6wwQIB2yANACYgDUAPgAIBMAjALABQDAzAARxkokXjgRwAxjF4BhXgG9OvJbzCYAHhEq9MJKWoDa3ALoBuTgF8gA===)

Reply:
From LDM, turns out the fixed-size buffer approach doesn't work for this scenario (GC tracking wouldn't work).
Fixed-size buffers avoid this today by being limited to simple struct types.
```
public unsafe struct C {
    fixed object x[3]; // error CS1663: Fixed size buffer type must be one of the following: bool, byte, short, int, long, char, sbyte, ushort, uint, ulong, float or double
}
```

Reply:
Is there any need for a struct at all? Why not just use locals directly?

Reply:
Locals are not required to be stored in a sequentual order - usually they are not even stored in memory but in CPU register.
I think we should not force a lowering strategy (e.g. stackalloc, custom type with seq. fields or fixed size buffer).
We should just mention those as possible implementations.